### DISCUSSION TITLE
Generate URL using route name

### DISCUSSION THREAD

**Discussion by scofield-ua** [nuxt/nuxt#17169] [Questions] [answered]
Hello. Is it possible to generate URL (including domain) by route name.

For example, in Laravel there is a function - `route('routeName')` which will generate URL basing on route name - `routeName` from routes files.

```
route('login') -> "http://localhost/login"
```

Maybe there is some "global" helper which can be used inside and outside of the template.

**Comment by Flanker72** [comment, marked as answer]
As I can see, there is no direct helper for this. Nuxt3 is based on Vue router, and Vue router resolves the URL regarding the app base.
To get full URL you can use this behavior:
```ts
const { resolve } = useRouter();
const route = resolve({ name: "login" });
const absoluteUrl = new URL(route.href, window.location.origin).href;
```
However, this can be done only on the client-side, SSR will fail because `window` object is not accessible.

**Reply by danielroe** [reply]
Plus, `ufo` has some useful utils to create a full URL if you need to do so on the server.

**Reply by scofield-ua** [reply]
Thanks for the tip. 

`window.location.origin` can be replaced with env variable.

Here is plugin I made for Nuxt 2:
```
export default ({ $config, app }, inject) => {
  const methods = {
    'route': (name, routeParams = {}) => {
      const query = routeParams.query === undefined ? {} : routeParams.query;
      const params = routeParams.params === undefined ? {} : routeParams.params;

      const route = app.router.resolve({
        name,
        query,
        params
      });

      const absoluteUrl = new URL(route.href, $config.appBaseUrl).href;

      return absoluteUrl;
    }
  }

  inject('url', methods);
}
```

And then you call it like this: `this.$url.route('login')` or pass `query` or [`params`](https://router.vuejs.org/api/#routelocationraw) with second argument.
